pkg/server/server_test.go

```diff
@@ -538,6 +538,65 @@ type repo
 	require.True(t, checkResponse.Allowed)
 }
 
+func TestCheckWithSingleflightResolution(t *testing.T) {
```

Comment:
Can we simplify this so it is more directly obvious of the side effects and less cognitive overhead?

I may suggest
```
type user

type folder
  relations
    define parent1: [folder]
    define parent2: [folder]
    define viewer: [user] or viewer from parent1 or viewer from parent2

- folder:x#viewer@user:jon
- folder:1#parent1@folder:x
- folder:1#parent2@folder:x

Check(folder:1#viewer@user:jon)
```

This is easier to comprehend at face value. We expect two dispatches of `Check(folder:x#viewer@user:jon)` to occur and thus one deduplicated subproblem.

Having the for...loop tuple generation below adds more cognitive overhead than we really need.

Reply:
Don't disagree that it's more overhead but I think it serves as a strong comparison of DB queries between with vs without singleflight resolver. This isn't deterministic so the for loop allows for a much higher probability that the test will pass. 

But in the spirit of your suggestion I've separated this out into a different case so that the base case is easier to follow.